x-pack/plugins/cases/server/connectors/cases/cases_oracle_service.test.ts

```diff
@@ -0,0 +1,283 @@
+
+  let service: CasesOracleService;
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+    service = new CasesOracleService({ unsecuredSavedObjectsClient, log: mockLogger });
+  });
+
+  describe('getRecordId', () => {
+    it('return the record ID correctly', async () => {
+      const ruleId = 'test-rule-id';
+      const spaceId = 'default';
+      const owner = 'cases';
+      const grouping = { 'host.ip': '0.0.0.1' };
+
+      const payload = `${ruleId}:${spaceId}:${owner}:${stringify(grouping)}`;
+      const hash = createHash('sha256');
+
+      hash.update(payload);
+
+      const hex = hash.digest('hex');
+
+      expect(service.getRecordId({ ruleId, spaceId, owner, grouping })).toEqual(hex);
+    });
+
+    it('sorts the grouping definition correctly', async () => {
+      const ruleId = 'test-rule-id';
+      const spaceId = 'default';
+      const owner = 'cases';
+      const grouping = { 'host.ip': '0.0.0.1', 'agent.id': '8a4f500d' };
+      const sortedGrouping = { 'agent.id': '8a4f500d', 'host.ip': '0.0.0.1' };
+
+      const payload = `${ruleId}:${spaceId}:${owner}:${stringify(sortedGrouping)}`;
+      const hash = createHash('sha256');
+
+      hash.update(payload);
+
+      const hex = hash.digest('hex');
+
+      expect(service.getRecordId({ ruleId, spaceId, owner, grouping })).toEqual(hex);
+    });
+
+    it('return the record ID correctly without grouping', async () => {
+      const ruleId = 'test-rule-id';
+      const spaceId = 'default';
+      const owner = 'cases';
+
+      const payload = `${ruleId}:${spaceId}:${owner}`;
+      const hash = createHash('sha256');
+
+      hash.update(payload);
+
+      const hex = hash.digest('hex');
+
+      expect(service.getRecordId({ ruleId, spaceId, owner })).toEqual(hex);
+    });
+
+    it('return the record ID correctly with empty grouping', async () => {
+      const ruleId = 'test-rule-id';
+      const spaceId = 'default';
+      const owner = 'cases';
+      const grouping = {};
+
+      const payload = `${ruleId}:${spaceId}:${owner}:${stringify(grouping)}`;
+      const hash = createHash('sha256');
+
+      hash.update(payload);
+
+      const hex = hash.digest('hex');
+
+      expect(service.getRecordId({ ruleId, spaceId, owner, grouping })).toEqual(hex);
+    });
+
+    it('return the record ID correctly without rule', async () => {
+      const spaceId = 'default';
+      const owner = 'cases';
+      const grouping = { 'host.ip': '0.0.0.1' };
+
+      const payload = `${spaceId}:${owner}:${stringify(grouping)}`;
+      const hash = createHash('sha256');
+
+      hash.update(payload);
+
+      const hex = hash.digest('hex');
+
+      expect(service.getRecordId({ spaceId, owner, grouping })).toEqual(hex);
+    });
+
+    it('throws an error when the ruleId and the grouping is missing', async () => {
+      const spaceId = 'default';
+      const owner = 'cases';
+
+      // @ts-expect-error: ruleId and grouping are omitted for testing
```

Comment:
Does `ExclusiveUnion<OracleKeyWithOptionalKey, OracleKeyWithOptionalGrouping>;` guarantee that either can be null but not both?

Reply:
This is correct.